src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +50,129 @@ public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private Dictionary<Tuple<string, string>, EventBase> EventsInProgress = new Dictionary<Tuple<string, string>, EventBase>();
+
+        private Dictionary<string, List<EventBase>> CompletedEvents = new Dictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            EventsInProgress[new Tuple<string, string>(requestId, eventToStart["event_name"])] = eventToStart;
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            Tuple<string, string> eventKey = new Tuple<string, string>(requestId, eventToStop["event_name"]);
+
+            // Locate the same name event in the EventsInProgress map
+            EventBase eventStarted = null;
+            if (EventsInProgress.ContainsKey(eventKey))
+            {
+                eventStarted = EventsInProgress[eventKey];
+            }
+
+            // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+            // was called without a corresponding StartEvent
+            if (null == eventStarted)
```

Comment:
This `if` condition, as well as the leading comments which describing its purpose, were ported from current Android telemetry PR. It looks plausible to me :->  That being said, when in doubt, we can always double check with the original author, so that if we need to change it, we change it consistently. Is that fair?